src/main/java/org/assertj/core/internal/Maps.java

```diff
@@ -441,27 +507,22 @@ public void assertHasSameSizeAs(AssertionInfo info, Map<?, ?> map, Map<?, ?> oth
     throw failures.failure(info, shouldContainExactly(actual, asList(entries), notFound, notExpected));
   }
 
-  private <K, V> void compareActualMapAndExpectedKeys(Map<K, V> actual, K[] keys, Set<K> notExpected, Set<K> notFound) {
-
-    Map<K, V> actualEntries = new LinkedHashMap<>(actual);
-    for (K key : keys) {
-      if (actualEntries.containsKey(key)) {
-        // this is an expected key
-        actualEntries.remove(key);
-      } else {
-        // this is a not found key
-        notFound.add(key);
-      }
+  private <V, K> Map<K, V> instantiate(Class<? extends Map> mapClass) {
+    try {
+      Constructor<?> constructor = mapClass.getConstructor();
+      return (Map<K, V>) constructor.newInstance();
+    } catch (Exception e) {
+      // default map but might not honor mapClass semantics (ex: CaseInsensitiveMap)
+      return new LinkedHashMap<>();
     }
-    // All remaining keys from actual copy are not expected entries.
-    notExpected.addAll(actualEntries.keySet());
   }
 
   private <K, V> void compareActualMapAndExpectedEntries(Map<K, V> actual, Entry<? extends K, ? extends V>[] entries,
                                                          Set<Entry<? extends K, ? extends V>> notExpected,
                                                          Set<Entry<? extends K, ? extends V>> notFound) {
     Map<K, V> expectedEntries = entriesToMap(entries);
-    Map<K, V> actualEntries = new LinkedHashMap<>(actual);
+    Map<K, V> actualEntries = instantiate(actual.getClass());
+    actualEntries.putAll(actual);
```

Comment:
Why not use `clone` here as well?

Reply:
This part is invoked only by `assertContainsExactly` that I still need to rework, once done it should disappear (and `clone` will be used)